common/kvstore/src/main/java/org/apache/spark/util/kvstore/RocksDBIterator.java

```diff
@@ -272,4 +283,37 @@ static int compare(byte[] a, byte[] b) {
     return a.length - b.length;
   }
 
+  static class ResourceCleaner implements Runnable {
+
+    private final RocksIterator rocksIterator;
+    private final RocksDB rocksDB;
+    private final AtomicBoolean status = new AtomicBoolean(true);
+
+    public ResourceCleaner(RocksIterator rocksIterator, RocksDB rocksDB) {
+      this.rocksIterator = rocksIterator;
+      this.rocksDB = rocksDB;
+    }
+
+    @Override
+    public void run() {
+      if (status.compareAndSet(true, false)) {
+        rocksDB.getIteratorTracker().forEach(ref -> {
```

Comment:
hmm... Why not just change the input arg type of the `rocksDB.notifyIteratorClosed` method...

Reply:
```

void notifyIteratorClosed(RocksIterator rocksIterator) {
    iteratorTracker.removeIf(ref -> {
      RocksDBIterator<?> rocksDBIterator = ref.get();
      return rocksDBIterator != null && rocksIterator.equals(rocksDBIterator.getRocksIterator());
    });
  }

```

How about?